### PR TITLE
🆕 Integrate Foundation Models Available VIA `timm`: `UNI`, `Prov-GigaPath`, `H-optimus-0`

### PR DESCRIPTION
**Pull request by GeorgeBatch**
Making a pull request as discussed in issue https://github.com/TissueImageAnalytics/tiatoolbox/issues/855

---
**Copied from the issue:**

I think it would be useful to integrate pre-trained foundation models from other labs into tiatoolbox.models.architecture.vanilla.py.

Currently, the `_get_architecture()` function allows the use of models from `torchvision.models`.

But another function `_get_timm_architecture()` could be made to incorporate foundation models which are available from `timm` with weights on HuggingFace Hub. All the models from `timm` that I've used require users to sign the licence agreement with the authors, so the licencing question seems to be solved itself since there is no way users will get access to the model weights just through Tiatoolbox without getting the access request approved by the authors first.

### CODE REVIEW

**Comment by GeorgeBatch** [comment]
Only added UNI and Prov-GigaPath for now. Will add more after initial comments.

I do not like that `TimmBackbone` pretty much repeats the `CNNBackbone`. The only difference is the absence of global average pooling in `TimmBackbone`.

**Comment by GeorgeBatch** [comment]
I found this file for testing: https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/tests/models/test_arch_vanilla.py

There might be a problem regarding memory and compute resources when running some of the larger feature extractors through GitHub actions, e.g. Prov-GigaPath needs a considerable amount of memory just to be loaded.

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:85]: weights are not required to load a model. These are provided in the `https://github.com/TissueImageAnalytics/tiatoolbox/blob/015652cc5c2357070592c8d46f5f2ff9a905a5c6/tiatoolbox/models/engine/patch_predictor.py#L231` or through `https://github.com/TissueImageAnalytics/tiatoolbox/blob/dd712ad173710bd27824e1141e34cb23f84792d1/tiatoolbox/data/pretrained_model.yaml` by default.

**Comment by shaneahmed** [comment]
> Only added UNI and Prov-GigaPath for now. Will add more after initial comments.
> 
> I do not like that `TimmBackbone` pretty much repeats the `CNNBackbone`. The only difference is the absence of global average pooling in `TimmBackbone`.

In that case, you can inherit `CNNBackbone` and just define the functions which change like we are doing here  https://github.com/TissueImageAnalytics/tiatoolbox/blob/e4e6f2249da8647a3f16b0e9a0616985b782b448/tiatoolbox/models/engine/patch_predictor.py#L25 for PatchPredictor

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [reply tiatoolbox/models/architecture/vanilla.py:85]: Removing `weights` should also fix the tests.

Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:85]: Are you suggesting to completely remove the `weights` argument from `_get_timm_architecture()`?

I am hardcoding the HuggingFace path for `timm` in `_get_timm_architecture()` as an initial version. But ideally users would be able to specify the path from where to load the `timm` model weights.

Comment (shaneahmed) [reply tiatoolbox/models/architecture/vanilla.py:85]: Adding `weights` argument would make it inconsistent with the rest of the architectures and the Engines API. We deliberately kept `weights` separate from architecture definition as it allows to import different weights in different pipelines in a flexible way. Hardcoding will not allow much flexibility. Once the architecture is loaded, loading weights shouldn't be a problem through Engines.
We are carefully reviewing this in #578 and #783.

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:112]: This should raise Error https://docs.astral.sh/ruff/rules/assert/

Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:112]: It did, should I ignore it with NOQA? Or remove?

Ignoring will allow to avoid unnecessary silent errors while creating the prov-gigapath model.

**Comment by shaneahmed** [comment]
I have updated this branch to make sure that tests pass on Ubuntu-24 before we merge it with develop.

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [reply tiatoolbox/models/architecture/vanilla.py:85]: Sorry, I confused it with the engines. I see you have made it similar to `_get_architecture()` https://github.com/TissueImageAnalytics/tiatoolbox/blob/7aa3ca3e3fb563ef89b7fae84e35b9eff803d548/tiatoolbox/models/architecture/vanilla.py#L20

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:109]:
```suggestion
    elif arch_name == "prov-gigapath" and timm.__version__ > "1.0.0":
```

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:119]:
```suggestion
        msg = f"Architecture {arch_name} not supported. If you are loading timm models, only timm version > `1.0.3` are supported." 
```

File: tiatoolbox/models/architecture/vanilla.py
Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py]: removed in commit [6ec9ca1](https://github.com/TissueImageAnalytics/tiatoolbox/pull/856/commits/6ec9ca14a1ff0f6a443668ce8fe083d3a2dbe9e8)

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:116]: This line is too long and causing the tests to fail. If you do `pre-commit install` in your development environment then it will test formatting before making a commit. You can also run `ruff` manually in your terminal before making a commit.
```suggestion
        msg = f"Architecture {arch_name} not supported." 
              f"If you are loading timm models, only timm version > `1.0.3` are supported."
```

Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:116]: Thanks! I will do it from now on.

**Comment by GeorgeBatch** [comment]
@shaneahmed, I added tests that revealed a complication.

I made two classes:
* `TimmModel` inheriting from `CNNModel`
* `TimmBackbone` inheriting from `CNNBackbone`

But to use `super().__init__()`, I need to pass a `backbone` parameter, which is restricted to the list in `_get_architecture()`, while I want to use a backbone specified in `_get_timm_architecture()`. So I cannot pass the backbone I want to `super().__init__()` because it triggers the assertion error inside `_get_architecture()`.

**Temporary fix** - pass "alexnet" backbone since it's the fastest to make: [40f7f95](https://github.com/TissueImageAnalytics/tiatoolbox/pull/856/commits/40f7f958765b063bb46ad8af7beb24dcde082dec)

Do you have suggestions on how to fix this properly?

File: tests/models/test_arch_vanilla.py
Comment (GeorgeBatch) [inline tests/models/test_arch_vanilla.py:73]: There is a discrepancy between my local tests and the GitHub actions tests when I run
`pytest -svvv tests/models/test_arch_vanilla.py`

I introduced `pretrained` flag in [deea74f](https://github.com/TissueImageAnalytics/tiatoolbox/pull/856/commits/deea74f0e5f9549590a736617c9f9dc5e76f125c)

I renamed the file with my HF token to mimic the situation on GitHub.

If I use `pretrained=True`, I get an error locally
```
E               huggingface_hub.utils._errors.GatedRepoError: 401 Client Error. (Request ID: Root=1-6710df61-738432cb51e75c6b39dd6329;8eb03fe3-9536-4777-8295-a54990368d2e)
E               
E               Cannot access gated repo for url https://huggingface.co/MahmoodLab/UNI/resolve/main/model.safetensors.
E               Access to model MahmoodLab/UNI is restricted. You must have access to it and be authenticated to access it. Please log in.
```

But if I use the `pretrained=False` locally, all tests pass.

On GitHub, even using `pretrained=False` results in a slightly different error. It is no longer about weights but the configuration json file.
```
E               huggingface_hub.errors.GatedRepoError: 401 Client Error. (Request ID: Root=1-6710d846-63696e761354213311810b36;300f8cc2-1e6b-405a-8c99-ebc4470c6615)
E               
E               Cannot access gated repo for url https://huggingface.co/MahmoodLab/UNI/resolve/main/config.json.
E               Access to model MahmoodLab/UNI is restricted. You must have access to it and be authenticated to access it. Please log in.
```

Comment (shaneahmed) [reply tests/models/test_arch_vanilla.py:74]: The reason we need tests is to make sure that we update the code to any changes in API in future. Can we find a timm model which does not require login? We do not need to test each and every model on GitHub actions. If we can verify the API works that should be fine.

Comment (GeorgeBatch) [reply tests/models/test_arch_vanilla.py:74]: I’ll add the smallest EfficientNet

`model = timm.create_model('efficientnet_b0', pretrained=True)`

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:326]:
```suggestion
class TimmModel(ModelABC):
```

As you are using ModelABC.__init__, you do not need CNNModel unless there is any other functionality you need.

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:388]:
```suggestion
class TimmBackbone(ModelABC):
```

As you are using ModelABC.__init__, you do not need CNNBackbone unless there is any other functionality you need.

Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:388]: There is the `infer_batch()` method `TimmBackbone` inherits from `CNNBackbone`.

File: tiatoolbox/models/architecture/vanilla.py
Comment (GeorgeBatch) [reply tiatoolbox/models/architecture/vanilla.py:326]: There are the `infer_batch()` and `_postproc()` methods, which `TimmModel` inherits from `CNNModel`.

**Comment by GeorgeBatch** [comment]
Currently, there are 2 things failing:

1. **DeepSource: Python** - because of `TimmBackbone(CNNBackbone)` and `TimmModel(CNNModel)` using the __init__ method of `ModelABC`.
    -  As a fix, I can copy and paste the `infer_batch` methods from `CNNBackbone` and `CNNModel` and inherit from `ModelABC`: `TimmBackbone(ModelABC)` and `TimmModel(ModelABC)`
    - But the proper fix would be to inheriting directly from `ModelABC` or even `nn.Module` after `infer_batch` becomes a standalone function: https://github.com/TissueImageAnalytics/tiatoolbox/pull/867.

2. **CodeCov** complains because we are not testing the `timm` models with restricted access: `UNI`, `Prov-GigaPath`, `H-Optimus-0`.

**Comment by shaneahmed** [comment]
> Currently, there are 2 things failing:
> 
> 1. **DeepSource: Python** - because `TimmBackbone(CNNBackbone)` and `TimmModel(CNNModel)` use the `__init__` method of `ModelABC`.
>    
>    * As a fix, I can copy and paste the `infer_batch` methods from `CNNBackbone` and `CNNModel` and inherit from `ModelABC`: `TimmBackbone(ModelABC)` and `TimmModel(ModelABC)`
>    * But the proper fix would be to inherit directly from `ModelABC` or even `nn.Module` after `infer_batch` becomes a standalone function: [🎨 Refactor `ModelABC` to Help Use Default Torch Models #867](https://github.com/TissueImageAnalytics/tiatoolbox/pull/867).
> 2. **CodeCov** complains because we are not testing the `timm` models with restricted access: `UNI`, `Prov-GigaPath`, `H-Optimus-0`.

Thanks @GeorgeBatch. Can you give me write access to your fork? I can then push some changes and try to fix these issues? Alternatively, I can download the files and send you the files which you can update.

**Comment by GeorgeBatch** [comment]
> Thanks @GeorgeBatch. Can you give me write access to your fork? I can then push some changes and try to fix these issues? Alternatively, I can download the files and send you the files which you can update.

@shaneahmed, sure, just invited you to be a collaborator my fork.

**Comment by shaneahmed** [comment]
> Currently, there are 2 things failing:
> 
> 1. **DeepSource: Python** - because `TimmBackbone(CNNBackbone)` and `TimmModel(CNNModel)` use the `__init__` method of `ModelABC`.
>    
>    * As a fix, I can copy and paste the `infer_batch` methods from `CNNBackbone` and `CNNModel` and inherit from `ModelABC`: `TimmBackbone(ModelABC)` and `TimmModel(ModelABC)`
>    * But the proper fix would be to inherit directly from `ModelABC` or even `nn.Module` after `infer_batch` becomes a standalone function: [🎨 Refactor `ModelABC` to Help Use Default Torch Models #867](https://github.com/TissueImageAnalytics/tiatoolbox/pull/867).
> 2. **CodeCov** complains because we are not testing the `timm` models with restricted access: `UNI`, `Prov-GigaPath`, `H-Optimus-0`.

I have made some changes which should help. For codecov, you can add `  # pragma: no cover` to   `UNI`, `Prov-GigaPath`, `H-Optimus-0`

**Comment by GeorgeBatch** [comment]
Great, thanks! The `timm` version requirement is only necessary for `prov-gigapath` model. The model can be created with an older version of `timm`, e.g. 0.9.8, but it throws an error when you try to pass an input tensor of the correct shape (batch_size, 3, 224, 224) into it: https://github.com/prov-gigapath/prov-gigapath/issues/2

The error is also very unhelpful and does not point to the `timm` version: `ValueError: not enough values to unpack (expected 3, got 2)`

**Comment by shaneahmed** [comment]
> @shaneahmed, Great, thanks! The `timm` version requirement is only necessary for `prov-gigapath` model. The model can be created with an older version of `timm`, e.g. 0.9.8, but it throws an error when you try to pass an input tensor of the correct shape (batch_size, 3, 224, 224) into it: [prov-gigapath/prov-gigapath#2](https://github.com/prov-gigapath/prov-gigapath/issues/2)
> 
> The error is also very unhelpful and does not point to the `timm` version: `ValueError: not enough values to unpack (expected 3, got 2)`

ok great. The main thing now left is to fix coverage for TimmBackbone and add a jupyter notebook as an example to show how this works.

File: tiatoolbox/models/architecture/vanilla.py
Comment (shaneahmed) [inline tiatoolbox/models/architecture/vanilla.py:138]:
```suggestion
```

**Comment by GeorgeBatch** [comment]
There are 3 more families of models that can be integrated, but they require their own files to be created: [Hibou -b and -L](https://github.com/GeorgeBatch/LC25000-clean/blob/main/source/feature_extraction/models/hibou.py), [Phikon v1 and v2](https://github.com/GeorgeBatch/LC25000-clean/blob/main/source/feature_extraction/models/owkin_phikon.py
), [Virchow v1 and v2](https://github.com/GeorgeBatch/LC25000-clean/blob/main/source/feature_extraction/models/virchow.py)

Should I add those files and call them from `_get_timm_network()`? 

Hibou requires to trust remote code when creating the model, which I do not really like.

**Comment by GeorgeBatch** [comment]
@shaneahmed, I think this branch is ready to be merged into `develop`.

I added code for saving a mask into the Masking Notebook, which currently does not show how to save the masks:
https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/examples/03-tissue-masking.ipynb

I also added a comment showing that TimmBackbone could be used as an alternative:
```python
model = CNNBackbone("resnet50")  # TimmBackbone(backbone="UNI", pretrained=True)
```
in both slide graph notebooks:
https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/examples/inference-pipelines/slide-graph.ipynb
https://github.com/TissueImageAnalytics/tiatoolbox/blob/develop/examples/full-pipelines/slide-graph.ipynb

**Review by shaneahmed — APPROVED** [review]
Thanks @GeorgeBatch This would be very helpful.

**Pull request merged**
